huaweicloud/services/ddm/resource_huaweicloud_ddm_instance.go

```diff
@@ -0,0 +1,931 @@
+	if err != nil {
+		return diag.Errorf("error waiting for instance (%s) to running: %s", d.Id(), err)
+	}
+	return nil
+}
+
+func updateInstanceFlavor(ctx context.Context, d *schema.ResourceData, cfg *config.Config, region string) diag.Diagnostics {
+	// updateInstanceFlavor: update DDM instance flavor
+	var (
+		updateInstanceFlavorHttpUrl = "v3/{project_id}/instances/{instance_id}/flavor"
+		updateInstanceProduct       = "ddm"
+	)
+	updateInstanceFlavorClient, err := cfg.NewServiceClient(updateInstanceProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DDM client: %s", err)
+	}
+
+	updateInstanceFlavorPath := updateInstanceFlavorClient.Endpoint + updateInstanceFlavorHttpUrl
+	updateInstanceFlavorPath = strings.ReplaceAll(updateInstanceFlavorPath, "{project_id}", updateInstanceFlavorClient.ProjectID)
+	updateInstanceFlavorPath = strings.ReplaceAll(updateInstanceFlavorPath, "{instance_id}", fmt.Sprintf("%v", d.Id()))
+
+	updateInstanceFlavorOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+	}
+
+	flavorId := utils.ValueIngoreEmpty(d.Get("flavor_id"))
+	engineId := utils.ValueIngoreEmpty(d.Get("engine_id"))
+	specCode, getSpecCodeErr := getSpecCodeByFlavorId(d, updateInstanceFlavorClient, flavorId.(string), engineId.(string))
+	if err != nil {
+		return getSpecCodeErr
+	}
+	updateInstanceFlavorOpt.JSONBody = utils.RemoveNil(buildUpdateInstanceFlavorBodyParams(d, specCode))
+	_, err = updateInstanceFlavorClient.Request("PUT", updateInstanceFlavorPath, &updateInstanceFlavorOpt)
+	if err != nil {
+		return diag.Errorf("error updating DdmInstance: %s", err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"RESIZING"},
+		Target:       []string{"RUNNING"},
+		Refresh:      ddmInstanceStatusRefreshFunc(d.Id(), region, cfg),
+		Timeout:      d.Timeout(schema.TimeoutUpdate),
+		Delay:        100 * time.Second,
+		PollInterval: 10 * time.Second,
+	}
+
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.Errorf("error waiting for instance (%s) to running: %s", d.Id(), err)
+	}
+	return nil
+}
+
+func getSpecCodeByFlavorId(d *schema.ResourceData, client *golangsdk.ServiceClient, flavorId, engineId string) (string, diag.Diagnostics) {
+	// getDdmFlavors: Query the List of DDM flavors
+	var (
+		getDdmFlavorsHttpUrl = "v2/{project_id}/flavors"
+	)
+
+	getDdmFlavorsPath := client.Endpoint + getDdmFlavorsHttpUrl
+	getDdmFlavorsPath = strings.ReplaceAll(getDdmFlavorsPath, "{project_id}", client.ProjectID)
+
+	getDdmFlavorsQueryParams := buildGetFlavorsQueryParams(engineId, 0)
+	getDdmFlavorsPath += getDdmFlavorsQueryParams
+	getInstanceOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+	}
+
+	for {
+		getDdmFlavorsResp, err := client.Request("GET", getDdmFlavorsPath, &getInstanceOpt)
+		if err != nil {
+			return "", common.CheckDeletedDiag(d, err, "error retrieving DdmFlavors")
```

Comment:
we should return the error message directly

Reply:
done